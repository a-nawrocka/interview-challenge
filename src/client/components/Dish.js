import React from "react";

class Dish extends React.Component {
	onSelected = () => {
		this.props.onSelected(this.props.dish);
	};

	onRemoved = () => {
		this.props.onRemoved(this.props.dish);
	};

	render() {
		let dishComponent;

		if (this.props.onSelected) {
			dishComponent = (
				<li className="item" onClick={this.onSelected}>
					<h2>{this.props.dish.name}</h2>
					<p>
						{this.props.dish.dietaries.map((dietary, index) => (
							<span key={index} className="dietary">
								{dietary}
							</span>
						))}
					</p>
				</li>
			);
		} else if (this.props.onRemoved) {
			dishComponent = (
				<li className="item">
					<h2>{this.props.dish.name}</h2>
					<p>
						{this.props.dish.dietaries.map((dietary, index) => (
							<span key={index} className="dietary">
								{dietary}
							</span>
						))}
					</p>
					<button className="remove-item" onClick={this.onRemoved}>
						x
					</button>
				</li>
			);
		}

		return dishComponent;
	}
}

export default Dish;
