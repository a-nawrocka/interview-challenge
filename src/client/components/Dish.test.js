import React from "react";
import renderer from "react-test-renderer";
import Dish from "./Dish";
import { shallow } from "enzyme";

describe("Given valid dish object and onSelected function", () => {
	it("display dish information without remove button", () => {
		const expectedDish = {
			id: 1,
			name: "tasty dish",
			dietaries: ["v", "ve", "df"],
		};

		const component = renderer.create(<Dish dish={expectedDish} onSelected={() => {}} />);

		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe("When dish is selected", () => {
		it("adds dish", () => {
			const expectedDish = {
				id: 1,
				name: "tasty dish",
				dietaries: ["v", "ve", "df"],
			};
			const mockOnSelected = jest.fn();

			const dishComponent = shallow(<Dish dish={expectedDish} onSelected={mockOnSelected} />);
			dishComponent.find("li").simulate("click");

			expect(mockOnSelected.mock.calls.length).toBe(1);
			expect(mockOnSelected.mock.calls[0][0]).toBe(expectedDish);
		});
	});
});

describe("Given valid dish object and onRemoved function", () => {
	it("display dish information with remove button", () => {
		const expectedDish = {
			id: 1,
			name: "tasty dish",
			dietaries: ["v", "ve", "df"],
		};

		const component = renderer.create(<Dish dish={expectedDish} onRemoved={() => {}} />);

		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe("When dish is removed", () => {
		it("removes dish", () => {
			const expectedDish = {
				id: 1,
				name: "tasty dish",
				dietaries: ["v", "ve", "df"],
			};
			const mockOnRemoved = jest.fn();

			const dishComponent = shallow(<Dish dish={expectedDish} onRemoved={mockOnRemoved} />);
			dishComponent.find("button").simulate("click");

			expect(mockOnRemoved.mock.calls.length).toBe(1);
			expect(mockOnRemoved.mock.calls[0][0]).toBe(expectedDish);
		});
	});
});
