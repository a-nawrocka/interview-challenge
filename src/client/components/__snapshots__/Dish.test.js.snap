// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given valid dish object and onRemoved function display dish information with remove button 1`] = `
<li
  className="item"
>
  <h2>
    tasty dish
  </h2>
  <p>
    <span
      className="dietary"
    >
      v
    </span>
    <span
      className="dietary"
    >
      ve
    </span>
    <span
      className="dietary"
    >
      df
    </span>
  </p>
  <button
    className="remove-item"
    onClick={[Function]}
  >
    x
  </button>
</li>
`;

exports[`Given valid dish object and onSelected function display dish information without remove button 1`] = `
<li
  className="item"
  onClick={[Function]}
>
  <h2>
    tasty dish
  </h2>
  <p>
    <span
      className="dietary"
    >
      v
    </span>
    <span
      className="dietary"
    >
      ve
    </span>
    <span
      className="dietary"
    >
      df
    </span>
  </p>
</li>
`;
